SourceX/DiabloUI/diabloui.cpp

```diff
@@ -576,14 +576,26 @@ BOOL UiCreatePlayerDescription(_uiheroinfo *info, DWORD mode, char *desc)
 
 void DrawArt(int screenX, int screenY, Art *art, int nFrame, DWORD drawW)
 {
-	BYTE *src = (BYTE *)art->data + (art->width * art->height * nFrame);
-	BYTE *dst = &gpBuffer[screenX + 64 + (screenY + SCREEN_Y) * BUFFER_WIDTH];
+	if (screenY >= SCREEN_WIDTH || screenX >= SCREEN_WIDTH)
+		return;
+
+	BYTE *__restrict src = (BYTE *)art->data + (art->width * art->height * nFrame);
+	BYTE *__restrict dst = &gpBuffer[screenX + 64 + (screenY + SCREEN_Y) * BUFFER_WIDTH];
 	drawW = drawW ? drawW : art->width;
 
-	for (DWORD i = 0; i < art->height && i + screenY < SCREEN_HEIGHT; i++, src += art->width, dst += BUFFER_WIDTH) {
-		for (DWORD j = 0; j < art->width && j + screenX < SCREEN_WIDTH; j++) {
-			if (j < drawW && (!art->masked || src[j] != art->mask))
-				dst[j] = src[j];
+	const DWORD iSize = std::min(art->height, static_cast<DWORD>(SCREEN_HEIGHT - screenY));
+	const DWORD jSize = std::min(drawW, std::min(art->width, static_cast<DWORD>(SCREEN_WIDTH - screenX)));
+
+	if (art->masked) {
+		for (DWORD i = 0; i < iSize; ++i, src += art->width, dst += BUFFER_WIDTH) {

```

Comment:
We prefer post rather then pre increment/decrement :)

```suggestion
		for (DWORD i = 0; i < iSize; i++, src += art->width, dst += BUFFER_WIDTH) {
```

Reply:
Done. The reason I prefer pre-increment is because it's faster for iterators so no need to have to different styles for iterators vs numbers.

Reply:
Afaik that hasn't been the case for compiler over the last few years :)

Reply:
It is faster for iterators in some cases, e.g. when constructor or destructor have side effects, or are defined in a separate compilation unit with no LTO, etc.